docs/design/kmp-nversions.md

```diff
@@ -0,0 +1,45 @@
+# nVersionCount support for Key Management Provider
+
+Author: Josh Duffney (@duffney)
+
+Tracked issues in scope:
+
+- https://github.com/ratify-project/ratify/issues/1751
+
+Proposal ref:
+
+- https://github.com/ratify-project/ratify/blob/dev/docs/proposals/Automated-Certificate-and-Key-Updates.md
+
+## Problem Statement
+
+In version 1.3.0 and earlier, Ratify does not support the nVersionCount parameter for Key Management Provider (KMP) resources. This means that when a certificate or key is rotated, Ratify updates the cache with the new version and removes the previous one, which may not suit all use cases.
+
+For instance, if a user needs to retain the last three versions of a certificate or key in the cache, Ratify cannot meet this requirement without manually adjusting the KMP resource for each new version.
+
+By supporting nVersionCount, Ratify would allow users to specify how many versions of a certificate or key should be kept in the cache, eliminating the need for manual updates to the KMP resource.
+
+## Proposed Solution
+
+To address this challenge, this proposal suggests adding support for the `maxVersionCount` parameter to the KMP resource in Ratify. This parameter will allow users to specify the number of versions of a certificate or key that should be retained in the cache.
+
+When a new version of a certificate or key is created, Ratify will check the `maxVersionCount` parameter to determine how many versions should be retained in the cache. If the number of versions exceeds the specified count, Ratify will remove the oldest version from the cache.
```

Comment:
Do we have a proposed default maxVersionCount? is 3 version a good default?

Reply:
> Do we have a proposed default maxVersionCount? is 3 version a good default?

@yizha1 proposal discussed this but I don't believe there was a final decision. From my research, `3` is the most common. And I think that it's a good default. I've seen as high as `10` mentioned but that only applies to customers who have stringent data retention policies.

Reply:
> 1.What will happen to the cached versions if we reconcile the KMP? 2.How will the `nVersionCount` sequential id looks like in cache structure and in the verifier result?

1. When a reconcile occurs the `GetCertificate` is triggered to check the status of the certificate. If the certificate is enabled the logic will continue and existing keys will be updated and new keys will be enabled. However, if the key status is disabled the entry will be removed from the cache. (Note: this isn't the current logic, but my proposed logic) Currently, reconcile updates existing keys and or adds new ones if they are enabled but does not remove disabled entries.

2. Right now the data structure is `[KMPMAPKEY][]*x509.Certificate` for certs and `[KMPMAPKEY]PublicKey` for keys. `KMPMAPKEY`'s struct is:

```go
type KMPMapKey struct {
	Name    string
	Version string
}
```

That structure is flattened to `[]*x509.Certificate` when passed to the verifiers. And because of that, as long as there is a function that can provider that flattened data to the verifier we can craft the kmp data structure however we'd like. My thinking is we can either leave the data structure as it is and just append the new versions of the cert or key to the kmp cache as match on the KMPMAPKEY to remove entries when needed. Or change the data structure to group versions of a certificate together. Something like this:

```go
type CertificateVersion struct {
    Version   int    // Version number
    CertData  []byte // Certificate data (PEM format, DER, etc.)
}

type Certificate struct {
    Name     string                         // Name or identifier of the certificate
    Versions map[int]CertificateVersion     // Map of version numbers to CertificateVersion
}

type CertificatesMap struct {
    Certificates map[string]*Certificate    // Map of certificate names to Certificate objects
}
```

My only hesitation with modifying the kmp data structure is I'm uncertain what value it adds if it is eventually flattened anyway. But, perhaps there is some usefulness if making the data structure easier to parse and or locate related certs. That said, I would love to hear other perspectives on this one. :)

Reply:
Discussed in community meeting: 1. we want to update the data structure to reflect keyvault/kmp structure, however we want to make sure new data structure is generic and extensible. ( GCP/AWS providers/Hashicrop/trust signer) 2. We need to confirm the behavior of disabled secrets, if content cannot be retrieved, we should pipe out from cache to follow security practice.

Reply:
> Do we have a proposed default maxVersionCount? is 3 version a good default?

discussed in community meeting, we should check the default behaviour for secretstore csidriver , 1. is there a max version , 2. what level is max version, per object or cr level